test/benchmark_utils/test_benchmark_utils.py

```diff
@@ -0,0 +1,829 @@
+            # Storing entries as f"{c} {fn}" rather than [c, fn] adds some work
+            # reviving the artifact, but it makes the json much easier to read.
+            match = pattern.search(cs)
+            assert match is not None
+            c, fn = match.groups()
+            data.append(benchmark_utils.FunctionCount(count=int(c), function=fn))
+
+        return benchmark_utils.FunctionCounts(
+            tuple(sorted(data, reverse=True)),
+            inclusive=inclusive)
+
+    baseline_inclusive = to_function_counts(artifacts["baseline_inclusive"], True)
+    baseline_exclusive = to_function_counts(artifacts["baseline_exclusive"], False)
+
+    stats_no_data = benchmark_utils.CallgrindStats(
+        benchmark_utils.TaskSpec("y = torch.ones(())", "pass"),
+        number_per_run=1000,
+        built_with_debug_symbols=True,
+        baseline_inclusive_stats=baseline_inclusive,
+        baseline_exclusive_stats=baseline_exclusive,
+        stmt_inclusive_stats=to_function_counts(artifacts["ones_no_data_inclusive"], True),
+        stmt_exclusive_stats=to_function_counts(artifacts["ones_no_data_exclusive"], False),
+    )
+
+    stats_with_data = benchmark_utils.CallgrindStats(
+        benchmark_utils.TaskSpec("y = torch.ones((1,))", "pass"),
+        number_per_run=1000,
+        built_with_debug_symbols=True,
+        baseline_inclusive_stats=baseline_inclusive,
+        baseline_exclusive_stats=baseline_exclusive,
+        stmt_inclusive_stats=to_function_counts(artifacts["ones_with_data_inclusive"], True),
+        stmt_exclusive_stats=to_function_counts(artifacts["ones_with_data_exclusive"], False),
+    )
+
+    return stats_no_data, stats_with_data
+
+
+class TestBenchmarkUtils(TestCase):
+    def test_timer(self):
+        timer = benchmark_utils.Timer(
+            stmt="torch.ones(())",
+        )
+        sample = timer.timeit(5).median
+        self.assertIsInstance(sample, float)
+
+        median = timer.blocked_autorange(min_run_time=0.01).median
+        self.assertIsInstance(median, float)
+
+        # We set a very high threshold to avoid flakiness in CI.
+        # The internal algorithm is tested in `test_adaptive_timer`
+        median = timer.adaptive_autorange(threshold=0.5).median
+
+        # Test that multi-line statements work properly.
+        median = benchmark_utils.Timer(
+            stmt="""
+                with torch.no_grad():
+                    y = x + 1""",
+            setup="""
+                x = torch.ones((1,), requires_grad=True)
+                for _ in range(5):
+                    x = x + 1.0""",
+        ).timeit(5).median
+        self.assertIsInstance(sample, float)
+
+    class _MockTimer:
+        _seed = 0
+
+        _timer_noise_level = 0.05
+        _timer_cost = 100e-9  # 100 ns
+
+        _function_noise_level = 0.05
+        _function_costs = (
+            ("pass", 8e-9),
+            ("cheap_fn()", 4e-6),
+            ("expensive_fn()", 20e-6),
+        )
```

Comment:
Can't you just define this as a dictionary in the first place?

Reply:
I'm always wary of adding mutable class fields, so it's just habit at this point.